keps/sig-storage/1855-csi-driver-service-account-token/README.md

```diff
@@ -0,0 +1,242 @@
+- [Proposal](#proposal)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [API Changes](#api-changes)
+  - [Example Workflow](#example-workflow)
+  - [Notes/Constraints/Caveats](#notesconstraintscaveats)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha-&gt;Beta](#alpha-beta)
+    - [Beta-&gt;GA](#beta-ga)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Scalability](#scalability)
+- [Alternatives](#alternatives)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes a way to obtain service account token for pods that the CSI
+drivers are mounting volumes for. Since these tokens are valid only for a
+limited period, this KEP will also give the CSI drivers an option to re-execute
+`NodePublishVolume` to mount volumes.
+
+## Motivation
+
+Currently, the only way that CSI drivers acquire service account tokens is to
+directly read the token in the file system. However, this approach has
+uncharming traits:
+
+1.  It will not work for csi drivers which run as a different non-root user than
+    the pods. See
+    [file permission section for service account token](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20180515-svcacct-token-volumes.md#file-permission).
+2.  CSI driver will have access to the secrets of pods that do not use it
+    because the CSI driver should have a `hostPath` volume for the `pods`
+    subdirectory to read the token.
+3.  The audience of the token is defaulted to kube apiserver.
+4.  The token is not guaranteed to be available (e.g.
+    `automountServiceAccountToken=false`).
+
+### User stories
+
+- HashiCorp Vault provider for secret store CSI driver requires service
+  account token of the pods they are mounting secrets at to authenticate to
+  Vaults. The provisioned secrets also have given TTL in Vault, so it is
+  necessary get tokens after the initial mount.
+- Cert manager CSI dirver will create CertificateRequests on behalf of the
+  pods.
+- Amazon EFS CSI driver wants the service account tokens of pods to exchange
+  for AWS credentials.
+
+## Proposal
+
+### Goals
+
+- Allow CSI driver to request audience-bounded service account tokens of pods
+  from kubelet to `NodePublishVolume`.
+- Provide an option to re-execute `NodePublishVolume` in a best-effort manner.
+
+### Non-Goals
+
+- Other CSI calls may not acquire pods' service account tokens via this
+  feature.
+- Failed re-execution of `NodePublishVolume` will not unmount volumes.
+
+### API Changes
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ... // existing fields
+
+    RequiresRemount *bool
+    ServiceAccountToken *ServiceAccountToken
+}
+
+// ServiceAccountToken contains parameters of a token.
+type ServiceAccountToken struct {
+    Audiences []string
```

Comment:
I think this should probably only allow a single audience. I'd like @liggitt to weigh in on the API review. I don't think this needs to block though.

Reply:
yes, from the security perspective, it is better to only allow a single audience. the use case to consider is secrets store csi driver which possibly has multiple providers. there are two ways:
1. currently the approach is to distribute one token with multiple audiences to the secrets store csi driver. 
2. another way would be distribute multiple tokens (each has single audience). 

FYI @seh

Reply:
My suggestion to allow multiple audiences was for parity with the _TokenRequest_ API, and to allow operators greater flexibility in talking to, say, more than one Vault authentication role, where they each \demand a different audience. Doing so would also allow changing [the configuration of such a role](https://www.vaultproject.io/api/auth/kubernetes#audience), and allowing multiple CSI driver daemons to play along through the transition.

Reply:
both ways can suffice the multi-audience in a csi driver scenario. take the secrets store csi driver as an example, the csi driver only needs to pass the single token with multiple audiences to all the providers, while if adopting the second approach, the csi driver needs to hardcode a mapping between the tokens from kubelet and the providers. the advantage i can think of of the second approach is every provider gets its own token but i am not sure how much security sugar we get here. seems to me that the convenience of the first approach outweighs the advantages that the second approach provides to us.

Reply:
> the csi driver only needs to pass the single token with multiple audiences to all the providers

passing replayable credentials which are considered valid by multiple providers to all the providers does not seem like a good idea

> allow operators greater flexibility in talking to, say, more than one Vault authentication role, where they each demand a different audience

I'm not very familiar with this... is this a scenario where you are presenting the token to a single vault server, using multiple audiences as some sort of scoping mechanism?

Reply:
> is this a scenario where you are presenting the token to a single vault server, using multiple audiences as some sort of scoping mechanism?

Yes. A Vault operator can mount [the Kubernetes auth method](https://www.vaultproject.io/docs/auth/kubernetes) any number of times. Within each mounted auth method, you can [create any number of _roles_](https://www.vaultproject.io/api/auth/kubernetes#create-role). Each such role defines [which service account names](https://www.vaultproject.io/api/auth/kubernetes#bound_service_account_names), within [which namespaces](https://www.vaultproject.io/api/auth/kubernetes#bound_service_account_namespaces), and bound to [which audience](https://www.vaultproject.io/api/auth/kubernetes#audience) (and only a single audience, questioned [here](https://github.com/terraform-providers/terraform-provider-vault/pull/601#discussion_r353856689) and [here](https://github.com/hashicorp/vault-plugin-auth-kubernetes/pull/70#discussion_r353859622), alas too late) to accept.

In my CSI driver, a pod can read multiple Vault [KV secrets](https://www.vaultproject.io/api-docs/secret/kv/kv-v2) and request multiple [X.509 certificates](https://www.vaultproject.io/api-docs/secret/pki), each involving authenticating as a different Vault role. Each of these roles could demand that the principal present a service account token bound to a different audience. If a given service account token is bound to multiple audiences—perhaps the union of what these Vault roles demand—then a single token could suffice for authenticating as each Vault role.

Reply:
> passing replayable credentials which are considered valid by multiple providers to all the providers does not seem like a good idea

agreed. when passing multiple tokens to the csi driver, in the case of secrets store csi driver, it is required to implement a mapping between the tokens and the providers. my current idea is to pass a map in `VolumeAttributes`:
```go
csi.storage.k8s.io/serviceAccount.tokens: {
  'audience-0': {
    'token': token-0,
    'expiry': expiry-0,
  },
  'audience-1': {
    'token': token-1,
    'expiry': expiry-1,
  },
  ...
}
```
regarding Vault roles which might have different audiences, the csi driver will pass all the tokens with audience in the format 'vault-*' to Vault provider. of course, it is a non-goal in this KEP but it is something that the csi driver needs to implement to accommodate the Vault case.